backend/app/middlewares/authmiddleware.py

```diff
@@ -0,0 +1,61 @@
+import os
+import firebase
+
+from firebase import credentials, auth
+from fastapi import FastAPI, Request, HTTPException, status, Depends, HTTPException, Security
+from fastapi.responses import JSONResponse
+from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
+from typing import List
+
+app = FastAPI()
+
+#use is_authorized_by_role to get info about user - NOT DONE
+#input is user (get info when accessing endpoint?)
+#output is user info found using token - token, role, etc.
+class get_user_info(is_authorized_by_role):
+    def __init__(self, roles: List[str]):
+        #self.roles = roles
+
+    def __call__(self, user: dict = Depends(verify_firebase_token)):
+        #role = user.get("role")
+        #if role not in self.roles:
+        #    raise HTTPException(
+        #        status_code=403,
+        #        detail=f"Access denied: role '{role}' not authorized"
+        #    )
+        #return user
+
+#middleware function
+#input allowed roles, endpoint
+#output: allowing/disallowing access to endpoint
+def role_based_access_control(allowed_roles: List[str]):
+    def middleware_decorator(endpoint: Callable):
+        async def wrapper(request: Request, *args, **kwargs):
+            roles = getattr(request.state, "user_roles", [])
```

Comment:
it's much much safer to get a user's roles from firebase or the db ourselves using the methods in the auth service, instead of sending it in the request

imagine someone malicious adding in roles to their requests that they don't have access to, which would let them access other roles that they haven't been assigned to in the db